File: Junior/CS101 - Abstract Data Types/PA1/Lex.java
//Justin Fortner
//ID#1481947
//jfortner
//CS101 PA1 Tantalo
//Lex.java

import java.io.*;
import java.util.Scanner;
   
public class Lex{
   public static void main(String[] args) throws IOException{
      Scanner in = null;
      PrintWriter out = null;
      String line = null;
      String[] token = null;
      int i, n, lineNumber = 0;

	  //Quit with message stderr is more or less 
	  //than two strings are given on the command line
      if(args.length < 2 || args.length > 2){
         System.err.println("stderr Usage: FileIO infile outfile");
         System.exit(1);
      }
      
	  //raken in input file and output file
      in = new Scanner(new File(args[0]));
      out = new PrintWriter(new FileWriter(args[1]));

	  //intitalizes string to be empty
	  String file = "";
	  //out.println("before while");
      //while there is another line in the file add it to the string
	  while(in.hasNextLine()){
         line = in.nextLine()+"\n";
		 file += line;
		 //out.println("this is file: " + file);
		 //out.println("in while");
      }
	  //out.println("after while");
	  out.println("file: " + file);
	  //split around the new line characters
	  //and put those strings into an array
	  String[] fileArray = file.split("\n");
	  
	  //out.println(fileArray.length);
	  
	  for(int j = 0; j < fileArray.length; j++){
		  out.println("fileArray: " + fileArray[j]);
	  }
	  //create a new list
	  List L = new List();
	  //out.println(fileArray.length);
	  //place the index of the first element of fileArray
	  //into the new list
	  L.append(0);
	  //move the cursor to the front of the new list
	  L.moveFront();
	  
	  //the first element of fileArray is already in the list so skip over it
	  //loop through the rest of fileArray
	  for(int j = 1; j < fileArray.length; j++){
		//out.println(fileArray[j]);
		//while the index is defined in the list L loop though it
		while(L.index()>= 0){
			//out.println("in while");
			//if the element in the fileArray comes before the element already in
			//the list L sort its index before that element in list L
			//if elements are equal sort before
			if(fileArray[j].compareTo(fileArray[L.get()])<0 || fileArray[j].compareTo(fileArray[L.get()])==0){
				L.insertBefore(j);
				break;
			}
			//if we get to the end of the list append the list to add the index of
			//the element to the end
			else if(L.index() == L.length()-1 && fileArray[j].compareTo(fileArray[L.get()])>0){
				L.append(j);
				break;
			}
			//if we havent found where it goes in the list yet move to the
			//next element in list L
			else{
				L.moveNext();
			}
		}
		//out.println("out while");
		//move the index to the front of list L so we can loop through it again
		L.moveFront();
	  }
	  
	  //out.println(L.toString());
	  //move to the front of list L that has sorted indicies of 
	  //elements in fileArray
	  L.moveFront();
	  //while we are in the list loop through it
	  while(L.index()>= 0){
		  //print the element in fileArray at the index given by list L
		  out.println(fileArray[L.get()]);
		  //move to the enxt element in the list L
		  L.moveNext();
	  }
	  
	  //close the files used in this program
      in.close();
      out.close();
   }
}

File: Junior/CS101 - Abstract Data Types/LAB/PA2/backup/Lex.c
//Justin Fortner
//ID#1481947
//jfortner
//CS101 PA2 Tantalo
//Lex.c

#include"List.h"
#include<stdio.h>
#include <stdlib.h>
#include<string.h>


#define MAX_LEN 1000
   
int main(int argc, char * argv[]){
	//initialize variables
	FILE *in, *out;
	char line[MAX_LEN];
	char* lineList[MAX_LEN];
	int count = 0;
	List L = newList();
	// check command line for correct number of arguments
	if( argc != 3 ){
		printf("Usage: %s <input file> <output file>\n", argv[0]);
		exit(1);
	}

	// open files for reading and writing 
	in = fopen(argv[1], "r");
	out = fopen(argv[2], "w");

	//while there is a line in the fine loop through the file
	while( fgets(line, MAX_LEN, in) != NULL)  {
		//allocate memory for array
		lineList[count] = calloc(strlen(line)+1, 1);
		//copy the line into the lineList 
		strcpy(lineList[count], line);
		//increment the counter for number of lines in the file
		count++;
	}
	//fprintf(out, "lineList: %s \n", lineList[0]);
	//fprintf(out, "count: %d \n", count);
	//for (int i = 0; i < count; i++){
		//fprintf(out, "lineList: %s \n", lineList[i]);
	//}
	
	//put a 0 into the list for the first index 
	append(L, 0);
	//move to the front of said list
	moveFront(L);
	
	//loop though the list from the second element in the list to the last
	for(int j = 1; j < count; j++){
		//while the index is defined in the list L loop though it
		while(index(L)>= 0){
			//if the element in the lineList comes before the element already in
			//the list L sort its index before that element in list L
			//if elements are equal sort before
			if(strcmp(lineList[j], lineList[get(L)])<0 || strcmp(lineList[j], lineList[get(L)])==0){
				insertBefore(L, j);
				break;
			}
			//if we get to the end of the list append the list to add the index of
			//the element to the end
			else if(index(L) == length(L)-1 && strcmp(lineList[j], lineList[get(L)])>0){
				append(L, j);
				break;
			}
			//if we havent found where it goes in the list yet move to the
			//next element in list L
			else{
				moveNext(L);
			}
		}
		//move the index to the front of list L so we can loop through it again
		moveFront(L);
	  }
	  
	  //move to the front of list L that has sorted indicies of 
	  //elements in lineList
	  moveFront(L);
	  //while we are in the list loop through it
	  while(index(L)>= 0){
		  //print the element in lineList at the index given by list L
		  fprintf(out, "%s", lineList[get(L)]);
		  //move to the enxt element in the list L
		  moveNext(L);
	  }
	
	//free the memory for the list
	freeList(&L);
	//close the files
	fclose(in);
	fclose(out);

	return(0);
}